File: ArithmeticSlices.java
/**
 * LeetCode 413 Arithmetic Slices
 */
class Solution {
    public int numberOfArithmeticSlices(int[] nums) {
        if (nums == null || nums.length < 3) {
            return 0;
        }
        int[] dp = new int[nums.length];
        int res = 0;
        int diff = nums[1] - nums[0];
        for (int i = 2; i < nums.length; i++) {
            if (nums[i] - nums[i - 1] == diff) {
                dp[i] = dp[i - 1] + 1;
                res += dp[i];
            } else {
                diff = nums[i] - nums[i - 1];
            }
        }
        return res;
    }
}

File: ArithmeticSlices.go
func numberOfArithmeticSlices(nums []int) int {
	if nums == nil || len(nums) < 3 {
		return 0
	}
	n := len(nums)
	dp := make([]int,n)
	for i := 2;i < n;i++ {
		if (nums[i] - nums[i - 1]) == (nums[i - 1] - nums[i - 2]) {
			dp[i] = dp[i - 1] + 1
		}
	}
	res := 0
	for _,v := range dp {
		res += v
	}
	return res
}
